src/vsr/replica.zig

```diff
@@ -440,77 +440,69 @@ pub fn ReplicaType(
-                    self.journal.dirty.set(slot);
-                    // Don't touch faulty — if it is set, we don't want to unset it. The WAL slot
-                    // may contain a corrupt version is this op, and we don't want to incorrectly
-                    // nack it. (This is why we do not call replace_header()/set_header_as_dirty()
-                    // here.)
                 }
-            }
-
-            // If we were a lagging backup that installed an SV but didn't finish fast-forwarding,
-            // the vsr_headers head op may be part of the checkpoint after this one.
-            maybe(vsr_headers.slice[0].op > self.op_checkpoint_trigger());
+            } else {
+                // Fall-through to choose op-head from vsr_headers.
+                //
+                // "Highest op from log_view in WAL" is not the correct choice for op-head when
+                // recovering with a durable DVC (though we still resort to this if there are no
+                // usable headers in the vsr_headers). It is possible that we started the view and
+                // finished some repair before updating our view_durable.
+                //
+                // To avoid special-casing this all over, we pretend this higher op doesn't
+                // exist. This is safe because we never prepared any ops in the view we joined just
+                // before the crash.
+                assert(self.log_view < self.view);
+                maybe(self.journal.op_maximum() > vsr_headers.slice[0].op);
+            }
+
+            // Try to use vsr_headers to update our head op.
+            // Loading headers prior to the head would risk the following scenario:
+            // 1. Replica A prepares[/commits] op X.
+            // 2. Replica A crashes.
+            // 3. Prepare X is corrupted in the WAL.
+            // 4. Replica A recovers. During `Replica.open()`, Replica A loads the header
+            //    for op `X - journal_slot_count` (same slot, prior wrap) from vsr_headers
+            //    into the journal.
+            // 5. Replica A participates in a view-change, but nacks[/does not include] op X.
+            // 6. Op X is truncated.
+            for (vsr_headers.slice) |*vsr_header| {
+                if (vsr.Headers.dvc_header_type(vsr_header) == .valid and
+                    vsr_header.op <= self.op_checkpoint_trigger() and
+                    (op_head == null or op_head.? <= vsr_header.op))
+                {
+                    op_head = vsr_header.op;
 
-            var op_head = for (vsr_headers.slice) |*header| {
-                if (vsr.Headers.dvc_header_type(header) != .valid) continue;
-                if (header.op <= self.op_checkpoint_trigger()) break header.op;
-            } else op_head: {
+                    if (!self.journal.has(vsr_header)) {
+                        self.journal.set_header_as_dirty(vsr_header);
+                    }
+                    break;
```

Comment:
Maybe 

```
                    if (!self.journal.has(vsr_header)) {
                        assert(op_head == null or op_head.? < vsr_header.op)
                        self.journal.set_header_as_dirty(vsr_header);
                    }
                   op_head = vsr_header.op;

```

?

To double-check that vsr header and journal header don't disagree

Reply:
:thinking: I think that assertion is actually not right though — if `op_head.? == vsr_header.op`, then we do want to `set_header_as_dirty` to replace the existing journal header with the one from vsr_headers. That is, our WAL has an old version of the op, but we serialized a view change with a newer header, but did not yet repair the WAL, then crashed.

Reply:
Ah, right, journal and  SV may actually diagree with each other!